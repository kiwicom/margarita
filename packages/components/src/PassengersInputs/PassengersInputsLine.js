// @flow

import * as React from 'react';
import { View } from 'react-native';
import { Text, Icon, Stepper, StyleSheet } from '@kiwicom/universal-components';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

type Props = {|
  +icon: string,
  +label: string,
  +value: number,
  +min?: number,
  +max?: number,
  +underline?: boolean,
  +onChange: (updateValue: number) => void,
|};

class PassengersInputsLine extends React.Component<Props> {
  handleDecrement = () => {
    this.props.onChange(-1);
  };

  handleIncrement = () => {
    this.props.onChange(1);
  };

  render() {
    const { icon, label, value, min, max, underline } = this.props;
    return (
      <View style={styles.line}>
        <View style={styles.icon}>
          <Icon
            name={icon}
            color={defaultTokens.colorIconSecondary}
            size="medium"
          />
        </View>
        <View style={[styles.counter, underline && styles.counterUnderline]}>
          <Text style={styles.label}>{label}</Text>
          <Stepper
            onDecrement={this.handleDecrement}
            onIncrement={this.handleIncrement}
            number={value}
            showNumber
            numberStyle={styles.stepperNumber}
            min={min}
            max={max}
          />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  line: {
    flexDirection: 'row',
    height: parseInt(defaultTokens.heightInputNormal, 10),
  },
  icon: {
    paddingHorizontal: parseInt(defaultTokens.spaceSmall, 10),
    justifyContent: 'center',
  },
  counter: {
    flex: 1,
    paddingEnd: parseInt(defaultTokens.spaceSmall, 10),
    flexDirection: 'row',
    alignItems: 'center',
  },
  counterUnderline: {
    borderBottomColor: defaultTokens.paletteInkLighter,
    borderBottomWidth: StyleSheet.hairlineWidth,
  },
  label: {
    flex: 1,
    fontSize: parseInt(defaultTokens.fontSizeTextLarge, 10),
    color: defaultTokens.colorTextAttention,
  },
  stepperNumber: {
    width: parseInt(defaultTokens.spaceXXXLarge, 10),
    textAlign: 'center',
    fontSize: parseInt(defaultTokens.fontSizeTextNormal, 10),
    color: defaultTokens.colorTextAttention,
  },
});

export default PassengersInputsLine;
