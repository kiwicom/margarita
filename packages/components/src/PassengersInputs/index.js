// @flow

export { default as PassengersInputs } from './PassengersInputs';
