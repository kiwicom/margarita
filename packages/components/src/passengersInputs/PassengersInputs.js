// @flow

import * as React from 'react';
import { View } from 'react-native';
import { Button, StyleSheet } from '@kiwicom/universal-components';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import Text from '../text/Text';
import type { Props, State } from './PassengersInputsTypes';
import PassengersInputsLine from './PassengersInputsLine';
import {
  updateAdults,
  updateInfants,
  updateBags,
  getMaxAdults,
  getMaxInfants,
  getMaxBags,
} from './PassengersInputsHelpers';

class PassengersInputs extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    /**
     * NOTE: Component is uncontrolled because parent form state should be updated
     * only after change is confirmed by the save button
     */
    const { adults, infants, bags } = this.props;
    this.state = {
      adults,
      infants,
      bags,
    };
  }

  handleChangeAdults = (value: number) => {
    this.setState(prevState => updateAdults(prevState, value));
  };

  handleChangeInfants = (value: number) => {
    this.setState(prevState => updateInfants(prevState, value));
  };

  handleChangeBags = (value: number) => {
    this.setState(prevState => updateBags(prevState, value));
  };

  handleSavePress = () => {
    this.props.onSavePress({ ...this.state });
  };

  render() {
    const { adults, infants, bags } = this.state;
    const { onClosePress } = this.props;
    return (
      <View>
        <Text weight="bold" size="large" style={styles.sectionLabel}>
          Passenger
        </Text>
        <PassengersInputsLine
          icon="passenger"
          label="Adults"
          value={adults}
          min={1}
          max={getMaxAdults(this.state)}
          onChange={this.handleChangeAdults}
          underline
        />
        <PassengersInputsLine
          icon="child-friendly"
          label="Infants"
          value={infants}
          min={0}
          max={getMaxInfants(this.state)}
          onChange={this.handleChangeInfants}
        />
        <Text weight="bold" size="large" style={styles.sectionLabel}>
          Bags
        </Text>
        <PassengersInputsLine
          icon="baggage-set"
          label="Checked Bags"
          value={bags}
          min={0}
          max={getMaxBags(this.state)}
          onChange={this.handleChangeBags}
        />
        <View style={styles.bottomButtons}>
          <View style={styles.buttonWrapper}>
            <Button onPress={onClosePress} type="secondary" label="Close" />
          </View>
          <View style={styles.buttonWrapper}>
            <Button
              onPress={this.handleSavePress}
              type="primary"
              label="Save"
            />
          </View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  sectionLabel: {
    color: defaultTokens.colorTextAttention,
    marginTop: parseInt(defaultTokens.spaceLarge, 10),
    marginBottom: parseInt(defaultTokens.spaceSmall, 10),
    paddingHorizontal: parseInt(defaultTokens.spaceMedium, 10),
  },
  bottomButtons: {
    flexDirection: 'row',
    marginTop: parseInt(defaultTokens.spaceMedium, 10),
    margin: parseInt(defaultTokens.spaceXXSmall, 10),
  },
  buttonWrapper: {
    flex: 1,
    margin: parseInt(defaultTokens.spaceXXSmall, 10),
  },
});

export default PassengersInputs;
