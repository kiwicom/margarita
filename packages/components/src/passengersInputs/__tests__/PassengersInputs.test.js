// @flow

import * as React from 'react';
import { shallow, render, fireEvent } from 'react-native-testing-library';

import PassengersInputs from '../PassengersInputs';
import {
  updateAdults,
  updateInfants,
  updateBags,
  getMaxAdults,
  getMaxInfants,
  getMaxBags,
} from '../PassengersInputsHelpers';

const testData0 = {
  adults: 7,
  infants: 2,
  bags: 14,
};
const testData1 = {
  adults: 2,
  infants: 1,
  bags: 1,
};
const onClose = jest.fn();
const onSave = jest.fn();

const { getByText } = render(
  <PassengersInputs
    {...testData0}
    onClosePress={onClose}
    onSavePress={onSave}
  />,
);

it('should execute onClosePress method', () => {
  fireEvent(getByText('Close'), 'press');
});

it('should execute onSavePress method with proper value', () => {
  fireEvent(getByText('Save'), 'press');
  expect(onSave).toHaveBeenCalledWith(testData0);
});

it('renders', () => {
  expect(
    shallow(
      <PassengersInputs
        adults={1}
        infants={0}
        bags={0}
        onClosePress={jest.fn()}
        onSavePress={jest.fn()}
      />,
    ),
  ).toMatchSnapshot();
});

test('updateAdults > should return correct value on increment', () => {
  expect(updateAdults(testData0, 1)).toMatchObject(testData0);
});

test('updateAdults > should return correct value on decrement', () => {
  expect(updateAdults(testData0, -1)).toMatchObject({
    adults: 6,
    infants: 2,
    bags: 12,
  });
});

test('updateInfants > should return correct value on increment', () => {
  expect(updateInfants(testData0, 1).infants).toBe(2);
});

test('updateInfants > should return correct value on decrement', () => {
  expect(updateInfants(testData0, -1).infants).toBe(1);
});

test('updateBags > should return correct value on increment', () => {
  expect(updateBags(testData0, 1).bags).toBe(14);
});

test('updateBags > should return correct value on decrement', () => {
  expect(updateBags(testData0, -1).bags).toBe(13);
});

test('getMaxAdults > should return correct max value', () => {
  expect(getMaxAdults(testData1)).toBe(8);
});

test('getMaxInfants > should return correct max value', () => {
  expect(getMaxInfants(testData1)).toBe(2);
});

test('getMaxBags > should return correct max value', () => {
  expect(getMaxBags(testData1)).toBe(4);
});
