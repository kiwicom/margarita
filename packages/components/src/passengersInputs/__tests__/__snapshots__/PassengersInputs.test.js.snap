// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
Object {
  "output": <Component>
    <Text
      size="large"
      style={
        Object {
          "color": "#171b1e",
          "marginBottom": 12,
          "marginTop": 24,
          "paddingHorizontal": 16,
        }
      }
      weight="bold"
    >
      Passenger
    </Text>
    <PassengersInputsLine
      icon="passenger"
      label="Adults"
      max={9}
      min={1}
      onChange={[Function]}
      underline={true}
      value={1}
    />
    <PassengersInputsLine
      icon="child-friendly"
      label="Infants"
      max={1}
      min={0}
      onChange={[Function]}
      value={0}
    />
    <Text
      size="large"
      style={
        Object {
          "color": "#171b1e",
          "marginBottom": 12,
          "marginTop": 24,
          "paddingHorizontal": 16,
        }
      }
      weight="bold"
    >
      Bags
    </Text>
    <PassengersInputsLine
      icon="baggage-set"
      label="Checked Bags"
      max={2}
      min={0}
      onChange={[Function]}
      value={0}
    />
    <Component
      style={
        Object {
          "flexDirection": "row",
          "margin": 4,
          "marginTop": 16,
        }
      }
    >
      <Component
        style={
          Object {
            "flex": 1,
            "margin": 4,
          }
        }
      >
        <Button
          onPress={[MockFunction]}
          type="secondary"
        >
          Close
        </Button>
      </Component>
      <Component
        style={
          Object {
            "flex": 1,
            "margin": 4,
          }
        }
      >
        <Button
          onPress={[Function]}
          type="primary"
        >
          Save
        </Button>
      </Component>
    </Component>
  </Component>,
}
`;
